doc/proposals/composable-integration-in sdk.md

```diff
@@ -0,0 +1,173 @@
+	Name               string   `json:"name,omitempty"`
+	Labels             []string `json:"labels,omitempty"`
+	Namespace          string   `json:"namespace,omitempty"`
+	Path               string   `json:"path"`
+	FormatTransformers []string `json:"format-transformers,omitempty"`
+}
+```
+
+An `ObjectRef` can be used to specify the type of any field of a CRD definition, allowing the value to be determined dynamically.
+For a detailed explanation of how to specify an object reference according to this schema, see [here](https://github.com/IBM/composable/blob/master/README.md#getvaluefrom-elements).
+
+The Composable SDK offers the following function for resolving the value of cross-resource references.
+
+```golang
+func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError 
+```
+
+The function `ResolveObject` takes a Kubernetes client and configuration, and `object` to resolve, and a blank object
+`resolved` that will contain the result, as well as a namespace. The namespace is the default used when references
+do not specify one. This function will cast the result to the type of the `resolved` object, provided that
+appropriate data transformers have been included in the reference definitions (see [tutorial](https://github.com/IBM/composable/blob/master/sdk/docs/tutorial.md) for an example).
+
+The `ResolveObject` function uses caching for looking up object kinds, as well as for the objects themselves, in order
+to ensure that a consistent view of the data is obtained. If any data is not available at the time of the lookup,
+it returns an error. So this function either resolves the entire object or it doesn't -- there are no partial results.
+The cache lasts only for the duration of the `ResolveObject` call.
+
+Our proposal is to have a tight integration of `ResolveObject` into the API offered by the Operator-SDK to interact with
+Kubernetes objects.
+
+The return value is a `ComposableError`:
+
+```golang
+type ComposableError struct {
+	Error error
+	// This indicates that the consuming Reconcile function should return this error
+	ShouldBeReturned bool
+}
+```
+
+The type `ComposableError` contains the error, if any, and a boolean `ShouldBeReturned` to indicate whether the calling reconciler should return this error or not. This is used to distinguish errors that are minor and could be fixed by immediately retrying from errors that may indicate a stronger failure, such as an ill-formed yaml (in which case there is no need to retry right away). If there are no errors, then the `ResolveObject` returns nil.
+
+
+
+
+### Risks and Mitigations
+
+Using the Composable SDK means that the CRD needs to have permission to access objects that are being referenced.
+This means that the operator developer needs to consider what kinds of objects will be permitted for references and
+add appropriate permissions for the RBAC rules of the CRD.
+
```

Comment:
Hi @vazirim, 

Really thank you for share your ideas and contribute with, first at all 🥇.

I have a few questions/concerns which follow. Please let me know if I misunderstood something. 

So, with this solution, in a CR of a CRD for example ( Memcached ), we can map many references for pre-exisent resources in the cluster and then, it will be used to find the data by the `ResolveObject`. Following code from the tutorial just to clarify my questions :-) 

```go
       // Resolve the memcached instance
	resolved := &cachev1alpha1.MemcachedResolved{}
	comperr := sdk.ResolveObject(r.client, r.config, memcached, resolved, request.Namespace)
	if comperr != nil {
		if comperr.ShouldBeReturned {
			return reconcile.Result{}, comperr.Error
		}
		reqLogger.Info("Error during resolve, but not returned", comperr.Error)
		return reconcile.Result{}, nil
	}
```

So, my main concerns/questions are:

- How it will deal with error scenarios, for example,  if 3 of 10 resources mapped for Memcached cannot be found? Will the Memcached resource be created or not? Has it a way to tell that some ref is optional?
- If the object, for example, a ConfigMap be found but it has not the key mapped in the CR how will it work?
- How would be the performance of it if the user adds too many resources refs to be "resolved"?
- How to validate the data in the CR? Has it already something implemented for that?
- Also, I think that would not be possible to call forever the reconcile in the case of something nothing works.
- And, in real scenarios, I am afraid that would not possible to solve it by changing the options in the manager to re-trigger the reconcile every few seconds as in the demo. It would need to have a watch/observer impl for each resource referenced. 

WDYT folks @estroz @joelanford @shawn-hurley  ?

Reply:
I would add one more, which is should there be some way to specify a default if it can not be resolved?

Reply:
Hi @camilamacedo86,

Thanks for your feedback!
- If 3 out of 10 resources cannot be found, then ResolveObject returns an error. It either resolves it all, or nothing. It also employs caching to get a consistent view. The cache is made fresh for each invocation of ResolveObject.

- If the object is found, but the data in it doesn't exist, then ResolveObject also returns an error.

- The performance hasn't been an issue so far because we have only seen CRs with a handful of references and no more. Resolving those references amounts to looking up in etcd. In any case, caching also alleviates performance.

- We have a validation that happens inside the Composable operator controller (so that the underlying template is well-formed). For the Composable-SDK, since the developer is modifying the CRD definition, and the open-api spec of the CRD, we get validation through that.

- The decision to retry is up to the consumer of the composable-SDK. But any operator will have a similar problem -- you need to decide for various error scenarios whether it makes sense to requeue another Reconcile cycle or not. So this is not specific to composable-SDK.

- Sure, having watchers works as well. In the operators we have developed, we like to resync periodically to check for the health of what the operator is managing.

--Mandana

Reply:
@shawn-hurley, yes we had thought of having defaults in the context of the Composable operator because it helped with validation (we could do a dry-run deployment). But we opted for not having defaults. The idea being that we want to have the ability of resolving data-dependencies dynamically. So if a resources has not been created yet, we retry until it's online. We have found this use case to be useful.

Reply:
@shawn-hurley The latest version of the Composable SDK (v.0.1.4) has support for default values.